From: SJ Park <sj@kernel.org>
Date: Thu, 23 Jul 2026 07:41:01 -0700
Subject: [PATCH] mm/damon/paddr: respect folio end for DAMOS_PAGEOUT

damon_pa_pageout() collects folios of the region by getting the folio of
region-internal addresses.  It iterates addresses starting from the
region start address and advancing the address by the size of the folio
of the address.  If the start address is in the middle of a large folio,
and if the next folios are small, some of the next folios could be
missed.  Fix the issue by advancing the address to exactly the start
address of the next folio.

The issue was discovered [1] by Sashiko.

[1] https://lore.kernel.org/20260517234112.89245-1-sj@kernel.org

Fixes: 3a06696305e7 ("mm/damon/ops: have damon_get_folio return folio even for tail pages")
Cc: <stable@vger.kernel.org> # 6.15.x
Signed-off-by: SJ Park <sj@kernel.org>
---
 mm/damon/paddr.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mm/damon/paddr.c b/mm/damon/paddr.c
index 8fb71183b76c0..6bb6e6b764ffb 100644
--- a/mm/damon/paddr.c
+++ b/mm/damon/paddr.c
@@ -304,7 +304,7 @@ static unsigned long damon_pa_pageout(struct damon_region *r,
 		else
 			list_add(&folio->lru, &folio_list);
 put_folio:
-		addr += folio_size(folio);
+		addr = PFN_PHYS(folio_pfn(folio)) + folio_size(folio);
 		folio_put(folio);
 	}
 	if (install_young_filter)
-- 
2.47.3

